src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/ConnectionTestWithSSLCert/CertificateTestWithTdsServer.cs

```diff
@@ -0,0 +1,146 @@
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using Microsoft.Data.SqlClient.ManualTesting.Tests.DataCommon;
+using Xunit;
+
+namespace Microsoft.Data.SqlClient.ManualTesting.Tests
+{
+    public class CertificateTestWithTdsServer
+    {
+        private static readonly string s_fullPathToPowershellScript = Path.Combine(Directory.GetCurrentDirectory(), "makepfxcert.ps1");
+        private static readonly string s_fullPathToPfx = Path.Combine(Directory.GetCurrentDirectory(), "localhostcert.pfx");
+        private static readonly string s_fullPathToClientCert = Path.Combine(Directory.GetCurrentDirectory(), "clientcert");
+
+        public CertificateTestWithTdsServer()
+        {
+            if (!Directory.Exists(s_fullPathToClientCert))
+            {
+                Directory.CreateDirectory(s_fullPathToClientCert);
+            }
+            CreatePfxCertificate(s_fullPathToPowershellScript);
+        }
+
+        [Theory]
+        [MemberData(nameof(ConnectionTestParametersData.GetConnectionTestParameters), MemberType = typeof(ConnectionTestParametersData))]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void BeginWindowsConnectionTest(ConnectionTestParameters connectionTestParameters)
+        {
+            ConnectionTest(connectionTestParameters);
+        }
+
+        [Theory]
+        [MemberData(nameof(ConnectionTestParametersData.GetConnectionTestParameters), MemberType = typeof(ConnectionTestParametersData))]
+        [PlatformSpecific(TestPlatforms.Linux)]
+        public void BeginLinuxConnectionTest(ConnectionTestParameters connectionTestParameters)
+        {
+            ConnectionTest(connectionTestParameters);
+        }
+
+        private void ConnectionTest(ConnectionTestParameters connectionTestParameters)
+        {
+            SqlConnectionStringBuilder builder = new(DataTestUtility.TCPConnectionString);
+
+            // The TestTdsServer does not validate the user name and password, so we can use any value if they are not defined.
+            string userId = string.IsNullOrWhiteSpace(builder.UserID) ? "user" : builder.UserID;
+            string password = string.IsNullOrWhiteSpace(builder.Password) ? "password" : builder.Password;
+
+            using TestTdsServer server = TestTdsServer.StartTestServer(enableFedAuth: false, enableLog: false, connectionTimeout: 15,
+                methodName: "", new X509Certificate2(s_fullPathToPfx, "nopassword", X509KeyStorageFlags.UserKeySet),
+                encryptionType: connectionTestParameters.TdsEncryptionType);
+
+            builder = new(server.ConnectionString)
+            {
+                UserID = userId,
+                Password = password,
+                TrustServerCertificate = connectionTestParameters.TrustServerCertificate,
+                Encrypt = connectionTestParameters.Encrypt,
+            };
+
+            if (connectionTestParameters.Certificate != string.Empty)
```

Comment:
Safer if you use below, this will avoid any potential NRE
```suggestion
            if (!string.IsNullOrEmpty(connectionTestParameters.Certificate))
```

Reply:
The suggestion to use safer method to avoid NRE was implemented.